syntax = "proto2";
package ncode.net;

// A graph link is a queue with a pipe attached to it. The
// queue has a service rate -- this is the speed of the link
message PBGraphLink {
  optional string src = 1;
  optional string dst = 2;
  optional uint64 bandwidth_bps = 4;
  optional uint32 src_port = 5;
  optional uint32 dst_port = 6;
  optional double delay_sec = 7;
}

// A region is a grouping of nodes.
message PBNetRegion {
  optional string id = 1;
  repeated string nodes = 2;
}

// The network.
message PBNet {
  repeated PBGraphLink links = 1; // at least one needed
  repeated PBNetRegion regions = 2;
}

// A 5-tuple that devices can match on.
message PBFiveTuple {
  optional uint32 ip_src = 1;
  optional uint32 ip_dst = 2;
  optional uint32 ip_proto = 3;
  optional uint32 src_port = 4;
  optional uint32 dst_port = 5;
}

message PBPath {
  repeated PBGraphLink links = 2;
}
